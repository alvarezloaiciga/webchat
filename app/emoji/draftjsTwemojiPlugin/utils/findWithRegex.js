const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
