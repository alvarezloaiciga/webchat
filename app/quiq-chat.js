'use strict';

Object.defineProperty(exports, '__esModule', {value: true});

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var fetch = _interopDefault(require('isomorphic-fetch'));
var uaParserJs = require('ua-parser-js');
var atmosphere = _interopDefault(require('atmosphere.js'));

function _toConsumableArray(arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) {
      arr2[i] = arr[i];
    }
    return arr2;
  } else {
    return Array.from(arr);
  }
}

//
/**
 * Formats the params object as query params in the url
 */
var formatQueryParams = function formatQueryParams(url, params) {
  var paramStrings = [];

  if (!params) {
    return url;
  }

  Object.keys(params).forEach(function(key) {
    var value = params[key];

    // If it's an array, we need to push once for each item
    if (Array.isArray(value)) {
      paramStrings.push.apply(
        paramStrings,
        _toConsumableArray(
          value.map(function(v) {
            return key + '=' + v;
          }),
        ),
      );
    } else if (value !== undefined) {
      paramStrings.push(key + '=' + value);
    }
  });

  if (paramStrings.length === 0) {
    return url;
  }

  return url + '?' + paramStrings.join('&');
};

var parser = new uaParserJs.UAParser();
var getBrowserName = function getBrowserName() {
  return parser.getResult().browser.name;
};
var getMajor = function getMajor() {
  return parseInt(parser.getResult().browser.major, 10);
};

var isIE9 = function isIE9() {
  return getBrowserName() === 'IE' && getMajor() <= 9;
};

//

var quiqChatSettings = void 0;

var setGlobals = function setGlobals(globals) {
  quiqChatSettings = globals;
};

var checkRequiredSettings = function checkRequiredSettings() {
  if (!quiqChatSettings || !quiqChatSettings.HOST || !quiqChatSettings.CONTACT_POINT) {
    throw new Error(
      '\n      HOST and CONTACT_POINT must be configured to call Quiq Messaging.\n      Did you forget to call init?\n      ',
    );
  }
};

var getContactPoint = function getContactPoint() {
  return quiqChatSettings.CONTACT_POINT;
};

var getPublicApiUrl = function getPublicApiUrl() {
  return quiqChatSettings.HOST + '/api/v1/messaging';
};

var getUrlForContactPoint = function getUrlForContactPoint() {
  return quiqChatSettings.HOST + '/api/v1/messaging/chat/' + quiqChatSettings.CONTACT_POINT;
};

//
var parseResponse = function parseResponse(response) {
  if (response.status && response.status >= 300) {
    return response
      .json()
      .then(function(res) {
        return Promise.reject(res);
      })
      .catch(function(err) {
        return Promise.reject(err);
      });
  }

  return response.json();
};

var joinChat$1 = function joinChat() {
  fetch(getUrlForContactPoint() + '/join', {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  });
};

var leaveChat$1 = function leaveChat() {
  fetch(getUrlForContactPoint() + '/leave', {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  });
};

var addMessage$1 = function addMessage(text) {
  fetch(getUrlForContactPoint() + '/send-message', {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({text: text}),
  });
};

var fetchWebsocketInfo = function fetchWebsocketInfo() {
  return fetch(getUrlForContactPoint() + '/socket-info', {
    mode: 'cors',
    credentials: 'include',
  }).then(parseResponse);
};

// Use socket-info as a ping since the ping endpoint isn't publicly exposed
var ping = function ping() {
  return fetchWebsocketInfo();
};

var fetchConversation$1 = function fetchConversation() {
  return fetch('' + getUrlForContactPoint(), {
    mode: 'cors',
    credentials: 'include',
  }).then(parseResponse);
};

var updateMessagePreview$1 = function updateMessagePreview(text, typing) {
  fetch(getUrlForContactPoint() + '/typing', {
    mode: 'cors',
    credentials: 'include',
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({text: text, typing: typing}),
  });
};

var checkForAgents$1 = function checkForAgents() {
  return fetch(
    formatQueryParams(getPublicApiUrl() + '/agents-available', {
      platform: 'Chat',
      contactPoint: getContactPoint(),
    }),
    {
      mode: 'cors',
      credentials: 'include',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    },
  ).then(parseResponse);
};

var _extends =
  Object.assign ||
  function(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };

//
var connection = void 0;
var callbacks = void 0;
var ie9Ping = void 0;

var buildRequest = function buildRequest(socketUrl) {
  // TODO: Add a way to specify transport
  var transport = 'websocket';
  if (isIE9()) {
    transport = 'jsonp';
  }

  /* eslint-disable no-use-before-define */
  return {
    url: 'https://' + socketUrl,
    enableXDR: true,
    withCredentials: true,
    contentType: 'application/json',
    logLevel: 'error',
    transport: transport,
    fallbackTransport: 'long-polling',
    trackMessageLength: true,
    maxReconnectOnClose: Number.MAX_SAFE_INTEGER,
    // Keep reconnectInterval at 10 seconds.  Otherwise if API-GW drops the atmosphere connection,
    // we will hammer them with onReconnect requests. See SER-2620
    reconnectInterval: 10000,
    onOpen: onOpen,
    onClose: onClose,
    onReopen: onReopen,
    onReconnect: onReconnect,
    onMessage: onMessage,
    onTransportFailure: onTransportFailure,
    onError: onError,
    onClientTimeout: onClientTimeout,
  };
  /* eslint-disable no-use-before-define */
};

var connectSocket = function connectSocket(builder) {
  var socketUrl = builder.socketUrl, websocketCallbacks = builder.callbacks;

  callbacks = websocketCallbacks;

  if (isIE9() && !ie9Ping) {
    // JSONP seems to be a bit unreliable, but we can prod it by periodically pinging the server...
    ie9Ping = setInterval(ping, 2000);
  }

  connection = _extends({}, atmosphere.subscribe(buildRequest(socketUrl)));
};

// ******** Atmosphere callbacks *********
var onOpen = function onOpen(response) {
  // Carry the UUID. This is required if you want to call subscribe(request) again.
  connection.request.uuid = response.request.uuid;
  callbacks.onConnectionEstablish && callbacks.onConnectionEstablish();
};

var onReopen = function onReopen() {
  callbacks.onConnectionEstablish && callbacks.onConnectionEstablish();
};

var onReconnect = function onReconnect(req) {
  // Long-polling doesn't clear up the error until it gets something back from the server
  // Force this to happen by sending a ping.
  clearTimeout(connection.pingTimeout);
  if (req.transport === 'long-polling') {
    connection.pingTimeout = setTimeout(function() {
      ping();
    }, connection.request.reconnectInterval + 5000);
  }
};

var onMessage = function onMessage(res) {
  var message = void 0;
  try {
    message = atmosphere.util.parseJSON(res.responseBody);
    callbacks.onMessage && callbacks.onMessage(message);
  } catch (e) {
    console.error('Error parsing Quiq websocket message');
    return;
  }
};

var onTransportFailure = function onTransportFailure(errorMsg, req) {
  callbacks.onTransportFailure && callbacks.onTransportFailure(errorMsg, req);
};

var onError = function onError() {
  callbacks.onConnectionLoss && callbacks.onConnectionLoss();
};

var onClientTimeout = function onClientTimeout() {
  callbacks.onConnectionLoss && callbacks.onConnectionLoss();
};

var onClose = function onClose() {
  callbacks.onClose && callbacks.onClose();
};

var _this = undefined;

function _asyncToGenerator(fn) {
  return function() {
    var gen = fn.apply(this, arguments);
    return new Promise(function(resolve, reject) {
      function step(key, arg) {
        try {
          var info = gen[key](arg);
          var value = info.value;
        } catch (error) {
          reject(error);
          return;
        }
        if (info.done) {
          resolve(value);
        } else {
          return Promise.resolve(value).then(
            function(value) {
              step('next', value);
            },
            function(err) {
              step('throw', err);
            },
          );
        }
      }
      return step('next');
    });
  };
}

//
var init = function init(settings) {
  var defaults = {
    CONTACT_POINT: 'default',
  };

  var globals = Object.assign({}, defaults, settings);

  setGlobals(globals);
};

var subscribe = (function() {
  var _ref = _asyncToGenerator(
    regeneratorRuntime.mark(function _callee(callbacks) {
      var wsInfo;
      return regeneratorRuntime.wrap(
        function _callee$(_context) {
          while (1) {
            switch ((_context.prev = _context.next)) {
              case 0:
                checkRequiredSettings();
                _context.next = 3;
                return fetchWebsocketInfo();

              case 3:
                wsInfo = _context.sent;

                connectSocket({socketUrl: wsInfo.url, callbacks: callbacks});

              case 5:
              case 'end':
                return _context.stop();
            }
          }
        },
        _callee,
        _this,
      );
    }),
  );

  return function subscribe(_x) {
    return _ref.apply(this, arguments);
  };
})();

var joinChat$$1 = function joinChat$$1() {
  checkRequiredSettings();
  return joinChat$1();
};

var leaveChat$$1 = function leaveChat$$1() {
  checkRequiredSettings();
  return leaveChat$1();
};

var addMessage$$1 = function addMessage$$1(text) {
  checkRequiredSettings();
  return addMessage$1(text);
};

var fetchConversation$$1 = function fetchConversation$$1() {
  checkRequiredSettings();
  return fetchConversation$1();
};

var updateMessagePreview$$1 = function updateMessagePreview$$1(text, typing) {
  checkRequiredSettings();
  return updateMessagePreview$1(text, typing);
};

var checkForAgents$$1 = function checkForAgents$$1() {
  checkRequiredSettings();
  return checkForAgents$1();
};

exports.init = init;
exports.subscribe = subscribe;
exports.joinChat = joinChat$$1;
exports.leaveChat = leaveChat$$1;
exports.addMessage = addMessage$$1;
exports.fetchConversation = fetchConversation$$1;
exports.updateMessagePreview = updateMessagePreview$$1;
exports.checkForAgents = checkForAgents$$1;
