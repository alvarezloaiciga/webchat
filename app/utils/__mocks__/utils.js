export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
