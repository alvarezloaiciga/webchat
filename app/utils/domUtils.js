import ReactDOM, {render} from 'react-dom';
import React from 'react';
import Routes from 'Routes';
import {quiqContainerId} from 'Common/Constants';
import {AppContainer} from 'react-hot-loader';
import {IntlProvider} from 'react-intl';
import Redbox from 'redbox-react';
import {Provider} from 'react-redux';

export const appIsMounted = (): boolean => {
  const container = document.getElementById(quiqContainerId);
  if (!container) return false;
  return container.childElementCount > 0;
};

export const destructApp = () => {
  ReactDOM.unmountComponentAtNode(document.getElementById(quiqContainerId));
};

export const constructApp = store => {
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <AppContainer errorReporter={Redbox}>
          <Routes />
        </AppContainer>
      </IntlProvider>
    </Provider>,
    document.getElementById(quiqContainerId),
  );

  if (module.hot) {
    module.hot.accept('Routes', () => {
      const NextApp = require('Routes').default; // eslint-disable-line global-require

      render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <AppContainer errorReporter={Redbox}>
              <NextApp />
            </AppContainer>
          </IntlProvider>
        </Provider>,
        document.getElementById(quiqContainerId),
      );
    });
  }
};

export const removeElement = (e: HTMLElement) => {
  e.parentNode.removeChild(e);
};

export const reloadApp = store => {
  destructApp();
  constructApp(store);
};

export const setFavicon = (iconUrl: string, appleTouchUrl: ?string) => {
  // Remove all icons
  const icons = [
    ...Array.from(document.querySelectorAll("link[rel='icon']")),
    ...Array.from(document.querySelectorAll("link[rel='apple-touch-icon']")),
  ];
  icons.forEach(icon => {
    removeElement(icon);
  });

  // Build icon
  const icon = document.createElement('link');
  icon.rel = 'icon';
  icon.href = iconUrl;

  // Build apple touch icon
  const appleTouchIcon = document.createElement('link');
  appleTouchIcon.rel = 'apple-touch-icon';
  appleTouchIcon.href = appleTouchUrl || iconUrl;

  // NOTE HARD: The apple-touch-icon must come BEFORE the regular icon.
  // Chrome looks for the last meta tag with a rel containing "icon" to try and use as title bar icon.
  const head = document.getElementsByTagName('head')[0];
  head.appendChild(appleTouchIcon);
  head.appendChild(icon);
};

export const setApplicationName = (name: string) => {
  const appName = document.querySelector("meta[name='application-name']");
  if (appName) {
    removeElement(appName);
  }
  const newAppName = document.createElement('meta');
  newAppName.name = 'application-name';
  newAppName.content = name;
  document.getElementsByTagName('head')[0].appendChild(newAppName);
};

export const setBrowserThemeColor = (color: string) => {
  const colorTag = document.querySelector("meta[name='theme-color']");
  if (colorTag) {
    removeElement(colorTag);
  }
  const newColor = document.createElement('meta');
  newColor.name = 'theme-color';
  newColor.content = color;
  document.getElementsByTagName('head')[0].appendChild(newColor);
};
