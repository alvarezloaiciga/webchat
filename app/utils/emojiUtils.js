import {emojiIndex} from 'emoji-mart';
import emojiRegexFactory from 'emoji-regex';
import invertBy from 'lodash/invertBy';
import type {EmojiMetadata, Emoji} from 'Common/Types';
import quiqOptions from 'Common/QuiqOptions';

const emojiRegex = emojiRegexFactory();

// Build lookup table of {unifiedCode : emojiId}
const emojiIdByUnified = {};
Object.keys(emojiIndex.emojis).forEach(id => {
  const unified = emojiIndex.emojis[id].unified.toLowerCase();
  emojiIdByUnified[unified] = id;
});

export const isSingleEmoji = (text: string): boolean => {
  // We know a string contains only one emoji iff
  // 1) An emoji regex finds only one match in the string, and
  // 2) The number of characters in that emoji === the number of characters in the string
  const emojis = text.match(emojiRegex);

  if (!emojis)
    return false;

  const emojiCharCount = emojis.reduce((count, e) => count + e.length, 0);
  return emojis.length === 1 && text.length === emojiCharCount;
};

export const convertUnicodeToEmojiObject = (u: string): ?Emoji => {
  const emojiId = Object.keys(emojiIndex.emojis).find(k => emojiIndex.emojis[k].native === u);
  return emojiIndex.emojis[emojiId];
};

// This function is used for filtering emojis in our picker, as well as redacting emojis prior to sending a message.
export const emojiFilter = (e: EmojiMetadata | Emoji | string) => {
  let emojiId;
  // emoji-mart will call this function when user is searching, but it only passes unicode string:
  if (typeof e === 'string') {
    emojiId =  emojiIdByUnified[e.toLowerCase()];
  } else {
    // NOTE: In emoji-mart land, id === short_codes[0]
    emojiId = e.id || e.short_names[0];
  }

  // Don't include any emojis that don't have id or shortcode (we don't know what they are)
  if (!emojiId)
    return false;

  const {includeEmojis, excludeEmojis} = quiqOptions;
  // Prioritize include over exclude. Only one of the should be defined by customer, however.
  if (Array.isArray(includeEmojis)) {
    return includeEmojis.includes(emojiId);
  }
  else if (Array.isArray(excludeEmojis)) {
    return !excludeEmojis.includes(emojiId);
  }

  // If neither exclude nor include was defined, always include emoji.
  return true;
};
