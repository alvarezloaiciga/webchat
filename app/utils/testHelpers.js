// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
