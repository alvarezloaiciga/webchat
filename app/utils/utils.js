// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
