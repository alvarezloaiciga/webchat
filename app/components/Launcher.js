// @flow
import React, {Component} from 'react';
import classnames from 'classnames';
import {supportsFlexbox, supportsSVG} from 'utils/utils';
import {joinChat, leaveChat} from 'network/chat';
import ChatContainer from './ChatContainer';
import ToggleChatButton from './ToggleChatButton';
import './styles/Launcher.scss';

class Launcher extends Component {
  state = {
    chatOpen: false,
  };

  toggleChat = () => {
    this.setState(
      prevState => ({chatOpen: !prevState.chatOpen}),
      () => {
        this.state.chatOpen ? joinChat() : leaveChat();
      },
    );
  };

  render() {
    const classNames = classnames('Launcher', {
      noFlexbox: !supportsFlexbox(),
      noSvg: !supportsSVG(),
    });
    return (
      <div className={classNames}>
        <ChatContainer hidden={!this.state.chatOpen} />
        <ToggleChatButton toggleChat={this.toggleChat} chatOpen={this.state.chatOpen} />
      </div>
    );
  }
}

export default Launcher;
