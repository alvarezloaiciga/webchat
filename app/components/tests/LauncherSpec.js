// @flow
jest.mock('utils/utils');
jest.mock('quiq-chat');
import QUIQ from 'utils/quiq';
import React from 'react';
import {Launcher} from '../Launcher';
import {shallow} from 'enzyme';
import {TestIntlObject, getMockMessage} from 'utils/testHelpers';
import type {ShallowWrapper} from 'enzyme';
import type {LauncherProps} from '../Launcher';
import {checkForAgents, fetchConversation} from 'quiq-chat';

jest.useFakeTimers();

describe('Launcher component', () => {
  let wrapper: ShallowWrapper;
  let render: () => void;
  let testProps: LauncherProps;
  const mockCheckForAgents = (checkForAgents: any);
  const mockFetchConversation = (fetchConversation: any);

  beforeEach(() => {
    render = () => {
      testProps = {
        intl: TestIntlObject,
      };
      wrapper = shallow(<Launcher {...testProps} />);
      wrapper.instance().componentDidMount();
    };
  });

  describe('after a minute', () => {
    const mockResponse = new Promise(resolve => resolve({available: true}));

    beforeEach(() => {
      mockCheckForAgents.mockReturnValue(mockResponse);
      render();
      jest.runTimersToTime(1000 * 60);
    });

    it('calls checkForAgents again', () => {
      expect(mockCheckForAgents.mock.calls.length).toBe(2);
    });
  });

  describe('when agents are available', () => {
    const mockResponse = new Promise(resolve => resolve({available: true}));
    const conversationResponse = Promise.resolve({id: 'testId', messages: []});

    beforeEach(() => {
      mockCheckForAgents.mockReturnValue(mockResponse);
      mockFetchConversation.mockReturnValue(conversationResponse);
      render();
    });

    it('renders', async () => {
      await mockResponse;
      await conversationResponse;
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    });

    describe('when not hidden', () => {
      beforeEach(() => {
        wrapper.setState({chatOpen: true});
      });

      it('renders the chat container', async () => {
        await mockResponse;
        await conversationResponse;
        wrapper.update();
        expect(wrapper.find('ChatContainer').length).toBe(1);
      });
    });
  });

  describe('auto pop for chat', () => {
    const mockResponse = new Promise(resolve => resolve({available: true}));
    const conversationResponse = Promise.resolve({id: 'testId', messages: []});

    beforeEach(() => {
      QUIQ.AUTO_POP_TIME = 200;
      render();
    });

    it("opens the chat even if the end user doesn't click on it", async () => {
      await mockResponse;
      await conversationResponse;
      jest.runTimersToTime(200);
      wrapper.update();
      expect(wrapper.find('ChatContainer').prop('hidden')).toBe(false);
    });
  });

  describe('when agents are not available', () => {
    const mockResponse = new Promise(resolve => resolve({available: false}));
    const conversationResponse = Promise.resolve({id: 'testId', messages: []});

    beforeEach(() => {
      mockCheckForAgents.mockReturnValue(mockResponse);
    });

    describe('when there is not a conversation already in progress', () => {
      beforeEach(() => {
        mockFetchConversation.mockReturnValue(conversationResponse);
        render();
        wrapper.setState({chatOpen: true});
      });

      it('renders a placeholder', async () => {
        await mockResponse;
        await conversationResponse;
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('when agents are not available', () => {
      describe('when there is a conversation already in progress', () => {
        const mockConversation = {
          id: 'testConversation',
          messages: [getMockMessage(), getMockMessage()],
        };

        beforeEach(() => {
          mockFetchConversation.mockReturnValue(Promise.resolve(mockConversation));
          render();
        });

        it('renders the chat', async () => {
          await mockResponse;
          wrapper.update();
          expect(wrapper).toMatchSnapshot();
        });

        describe('after the chat has been closed', () => {
          beforeEach(() => {
            const closedConversation = Object.assign({}, mockConversation, {
              messages: [...mockConversation.messages, getMockMessage(3, {type: 'Leave'})],
            });
            mockFetchConversation.mockReturnValue(Promise.resolve(closedConversation));
            render();
            wrapper.instance().componentDidMount();
          });

          it('leaves the chat closed', async () => {
            await mockResponse;
            wrapper.update();
            expect(wrapper).toMatchSnapshot();
          });
        });
      });
    });
  });
});
