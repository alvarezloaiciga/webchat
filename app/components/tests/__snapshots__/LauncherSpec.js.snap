// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launcher component rendering renders 1`] = `
<div
  className="Launcher"
>
  <ChatContainer
    hidden={true}
  />
  <ToggleChatButton
    chatOpen={false}
    toggleChat={[Function]}
  />
</div>
`;
